libcudacxx/include/cuda/std/inplace_vector

```diff
@@ -0,0 +1,1808 @@
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr void resize(const size_type __count, const _Tp& __value)
+  {
+    const auto __diff = static_cast<ptrdiff_t>(__count) - static_cast<ptrdiff_t>(this->size());
+    if (__diff == 0)
+    {
+      return;
+    }
+    else if (__diff < 0)
+    {
+      this->__destroy(this->begin() + __count, this->end());
+    }
+    else
+    {
+      if (_Capacity < __count)
+      {
+        _CUDA_VSTD::__throw_bad_alloc();
+      }
+
+      const iterator __end = this->end();
+      this->__uninitialized_fill(__end, __end + __diff, __value);
+    }
+  }
+
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr void reserve(const size_type __count)
+  {
+    if (_Capacity < __count)
+    {
+      _CUDA_VSTD::__throw_bad_alloc();
+    }
+  }
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr void shrink_to_fit() noexcept {}
+
+  _LIBCUDACXX_TEMPLATE(class _Tp2 = _Tp)
+  _LIBCUDACXX_REQUIRES(_CCCL_TRAIT(is_swappable, _Tp2) _LIBCUDACXX_AND _CCCL_TRAIT(is_move_constructible, _Tp2))
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr void swap(inplace_vector& __other) noexcept(
+    _Capacity == 0 || (_CCCL_TRAIT(is_nothrow_swappable, _Tp2) && _CCCL_TRAIT(is_nothrow_move_constructible, _Tp2)))
+  {
+    _CCCL_IF_CONSTEXPR (_Capacity == 0)
+    {
+      return;
+    }
+
+    if (this->size() < __other.size())
+    {
+      const auto __new_mid = _CUDA_VSTD::swap_ranges(this->begin(), this->end(), __other.begin());
+      this->__uninitialized_move(__new_mid, __other.end(), this->end());
+      __other.__destroy(__new_mid, __other.end());
+    }
+    else
+    {
+      const auto __new_mid = _CUDA_VSTD::swap_ranges(__other.begin(), __other.end(), this->begin());
+      __other.__uninitialized_move(__new_mid, this->end(), __other.end());
+      this->__destroy(__new_mid, this->end());
+    }
+  }
+
+  _LIBCUDACXX_TEMPLATE(class _Tp2 = _Tp)
+  _LIBCUDACXX_REQUIRES(_CCCL_TRAIT(is_swappable, _Tp2) _LIBCUDACXX_AND _CCCL_TRAIT(is_move_constructible, _Tp2))
+  _LIBCUDACXX_INLINE_VISIBILITY friend constexpr void swap(inplace_vector& __lhs, inplace_vector& __rhs) noexcept(
+    _Capacity == 0 || (_CCCL_TRAIT(is_nothrow_swappable, _Tp2) && _CCCL_TRAIT(is_nothrow_move_constructible, _Tp2)))
+  {
+    __lhs.swap(__rhs);
+  }
+
+  // inplace_vector.comparison
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_INLINE_VISIBILITY constexpr bool
+  operator==(const inplace_vector& __lhs, const inplace_vector& __rhs) noexcept(
+    noexcept(_CUDA_VSTD::equal(__lhs.begin(), __lhs.end(), __rhs.begin(), __rhs.end())))
+  {
+    return _CUDA_VSTD::equal(__lhs.begin(), __lhs.end(), __rhs.begin(), __rhs.end());
+  }
+#  if _CCCL_STD_VER <= 2017
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_INLINE_VISIBILITY constexpr bool
+  operator!=(const inplace_vector& __lhs, const inplace_vector& __rhs) noexcept(
+    noexcept(_CUDA_VSTD::equal(__lhs.begin(), __lhs.end(), __rhs.begin(), __rhs.end())))
+  {
+    return !_CUDA_VSTD::equal(__lhs.begin(), __lhs.end(), __rhs.begin(), __rhs.end());
+  }
```

Comment:
Suggestion: simplification:
```suggestion
  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_INLINE_VISIBILITY constexpr bool
  operator!=(const inplace_vector& __lhs, const inplace_vector& __rhs) noexcept(
    noexcept(!(__lhs == __rhs)))
  {
    return !(__lhs == __rhs);
  }
```

Reply:
Hmm... the main crux to me is that `lexicographical_compare` and `equal` are never `noexcept` in libcudacxx now. So the exception specification strengthening doesn't work at this moment.

Reply:
There are two things here. Yes the noexcept branch is correct, but on the other hand it adds one more step the user has to go through for debugging. 

So in case of "trivial" implementations like this, I am favoring open coding it whenever possible